Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean

```diff
@@ -0,0 +1,199 @@
+variable (R)
+
+lemma isSeparating : IsSeparating (freeYoneda R) := by
+  intro M N f₁ f₂ h
+  ext ⟨X⟩ m
+  obtain ⟨g, rfl⟩ := freeYonedaEquiv.surjective m
+  exact congr_arg freeYonedaEquiv (h _ ⟨X, rfl⟩ g)
+
+lemma isDetecting : IsDetecting (freeYoneda R) :=
+  (isSeparating R).isDetecting
+
+end freeYoneda
+
+instance wellPowered : WellPowered (PresheafOfModules.{u} R₀) :=
+  wellPowered_of_isDetecting (freeYoneda.isDetecting R₀)
+
+/-- The type of elements of a presheaf of modules. -/
+abbrev Elements {C : Type u₁} [Category.{v₁} C] {R : Cᵒᵖ ⥤ RingCat.{u}}
+  (M : PresheafOfModules.{v} R) := ((toPresheaf R).obj M ⋙ forget Ab).Elements
+
+/-- Given a presheaf of modules `M`, this is a constructor for the type `M.Elements`. -/
+abbrev elementsMk {C : Type u₁} [Category.{v₁} C] {R : Cᵒᵖ ⥤ RingCat.{u}}
+    (M : PresheafOfModules.{v} R) (X : Cᵒᵖ) (x : M.obj X) : M.Elements :=
+  Functor.elementsMk _ X x
+
+namespace Elements
+
+variable {C : Type u} [Category.{v} C] {R : Cᵒᵖ ⥤ RingCat.{v}} {M : PresheafOfModules.{v} R}
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is
+free presheaf of modules on the Yoneda presheaf of types corresponding to the
+underlying object of `m`. -/
+noncomputable abbrev freeYoneda (m : M.Elements) :
+    PresheafOfModules.{v} R := (free R).obj (yoneda.obj m.1.unop)
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is
+the canonical morphism `m.freeYoneda ⟶ M`. -/
+noncomputable abbrev fromFreeYoneda (m : M.Elements) :
+    m.freeYoneda ⟶ M :=
+  freeYonedaEquiv.symm m.2
+
+lemma fromFreeYoneda_app_apply (m : M.Elements) :
+    m.fromFreeYoneda.app m.1 (ModuleCat.freeMk (𝟙 _)) = m.2 := by
+  apply freeYonedaEquiv_symm_app
+
+end Elements
+
+section
+
+variable {C : Type u} [SmallCategory.{u} C] {R : Cᵒᵖ ⥤ RingCat.{u}} (M : PresheafOfModules.{u} R)
+
+/-- Given a presheaf of modules `M`, this is the coproduct of
+all free Yoneda presheaves `m.freeYoneda` for all `m : M.Elements`. -/
+noncomputable abbrev freeYonedaCoproduct : PresheafOfModules.{u} R :=
+  ∐ (Elements.freeYoneda (M := M))
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is the
+canonical inclusion `e.freeYoneda ⟶ M.freeYonedaCoproduct`. -/
+noncomputable abbrev ιFreeYonedaCoproduct (m : M.Elements) :
+    m.freeYoneda ⟶ M.freeYonedaCoproduct :=
+  Sigma.ι _ m
+
+/-- Given a presheaf of modules `M`, this is the
+canonical morphism `M.freeYonedaCoproduct ⟶ M`. -/
+noncomputable def fromFreeYonedaCoproduct :
+    M.freeYonedaCoproduct ⟶ M :=
+  Sigma.desc Elements.fromFreeYoneda
+
+/-- Given an element `m` of a presheaf of modules `M`, this is the associated
+canonical section of the presheaf `M.freeYonedaCoproduct` over the object `m.1`. -/
+noncomputable def freeYonedaCoproductMk (m : M.Elements) :
+    M.freeYonedaCoproduct.obj m.1 :=
+  (M.ιFreeYonedaCoproduct m).app _ (ModuleCat.freeMk (𝟙 _))
+
+@[reassoc (attr := simp)]
+lemma ι_fromFreeYonedaCoproduct (e : M.Elements) :
```

Comment:
Is there a reason you've switched notation from `m` to `e`? I don't have a preference but my instinct would be to be consistent.

Reply:
Thanks. I have fixed these inconsistencies.